storage/test/sparse_merkle/tree_test.cpp

```diff
@@ -273,13 +406,658 @@ TEST(tree_tests, multiple_updates) {
+  updates.emplace(Sliver("key191"), Sliver("val3"));
+  auto batch = tree.update(updates);
+
+  // There should be 3 internal nodes, with no leaf children in the root node, 1
+  // leaf child in the second node, and 2 leaf children in the 3rd node.
+  ASSERT_EQ(3, batch.internal_nodes.size());
+  ASSERT_EQ(0, batch.internal_nodes[0].second.numLeafChildren());
+  ASSERT_EQ(1, batch.internal_nodes[1].second.numLeafChildren());
+  ASSERT_EQ(2, batch.internal_nodes[2].second.numLeafChildren());
+  db_put(db, batch);
+
+  // Delete the LeafChild in the node at depth1. This should not trigger any promotion.
+  KeysVector deletes{Sliver("key9")};
+  batch = tree.remove(deletes);
+
+  // Only the first two nodes have been updated.
+  ASSERT_EQ(2, batch.internal_nodes.size());
+  ASSERT_TRUE(internalNodeVersionMatches(2, batch.internal_nodes));
+  auto& [root_key, root_node] = batch.internal_nodes[0];
+  auto& [depth1_key, depth1_node] = batch.internal_nodes[1];
+  ASSERT_EQ("", root_key.path().toString());
+  ASSERT_EQ("b", depth1_key.path().toString());
+  ASSERT_TRUE(validLinkExists(root_key, root_node, depth1_key, depth1_node));
+
+  // There are no leaf updates.
+  ASSERT_EQ(0, batch.leaf_nodes.size());
+
+  // The root node still has zero leaf children and the node at depth one has
+  // zero leaf children.
+  //
+  // Note that the version number of these children is still 1, as they were
+  // written in the initial update. The versions of the internal children above
+  // them, including the root of the BatchedInternalnode, have all had their
+  // versions bumped to indicate the update.
+  ASSERT_EQ(0, root_node.numLeafChildren());
+  ASSERT_EQ(0, depth1_node.numLeafChildren());
+
+  // There are 2 old internal keys (root, depth 1) and one stale leaf
+  ASSERT_EQ(Version(2), batch.stale.stale_since_version);
+  ASSERT_EQ(2, batch.stale.internal_keys.size());
+  ASSERT_TRUE(internalKeyExists("", 1, batch.stale.internal_keys));
+  ASSERT_TRUE(internalKeyExists("b", 1, batch.stale.internal_keys));
+  ASSERT_EQ(1, batch.stale.leaf_keys.size());
+  ASSERT_TRUE(leafKeyExists("key9", 1, batch.stale.leaf_keys));
+}
+
+// Add 4 nodes that overlap in the first nibble, with 2 nodes also matching in
+// one second nibble, and 2 nodes matching in the other second nibble. Thus the
+// tree of BatchedInternalNodes will look like the following:
+//
+//                Root
+//                 |
+//                "b"
+//                 |
+//           -------------
+//           |           |
+//         "ba"        "be"
+//
+// Each of "ba" and "be" contain 2 leaf children. Deleting a leaf child from "be"
+// should cause the other leaf child to go to "b", but leave "ba" alone. Thus
+// after the delete, the tree should look like the following:
+//
+//                Root
+//                 |
+//                "b"
+//                 |
+//                "ba"
+//
+TEST(tree_tests, remove_left_node_and_promote_leaving_right_node_alone) {
```

Comment:
This test provides a brilliant example, thank you!

Reply:
:)